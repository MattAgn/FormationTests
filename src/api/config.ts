export const EMAIL_API_ENDPOINT = 'https://staging.inshallah.com/auth/send-validation-email';
