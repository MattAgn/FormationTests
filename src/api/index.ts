import wretch from 'wretch';

export interface IAuthTokens {
  authToken: string;
  refreshToken: string;
}

export class BaseApi {
  public static baseCall() {
    return wretch('https://staging.inshallah.com/');
  }
}
