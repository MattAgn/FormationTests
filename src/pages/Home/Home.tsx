import React, { useState } from 'react';
import { View, Text, TextInput } from 'react-native';
import styled from 'styled-components/native';
import wretch from 'wretch';
import { Button } from '../../components/Button';

export const HomeBasic = () => {
  const [email, setEmail] = useState('');
  const [isSuccess, setIsSuccess] = useState<boolean | null>(null);

  const onValidate = () => {
    wretch('https://staging.inshallah.com/auth/send-validation-email')
      .post({ email })
      .res()
      .then(() => setIsSuccess(true))
      .catch(() => setIsSuccess(false));
  };

  return (
    <Container>
      <Title>Subscribe to our newsletter</Title>
      <TextInput placeholder="Email" value={email} onChangeText={setEmail} />
      <Button title="Validate" onPress={onValidate} />
      {isSuccess && <SuccessMessage>Email sent !</SuccessMessage>}
      {isSuccess === false && <ErrorMessage>Oups something went wrong...</ErrorMessage>}
    </Container>
  );
};

const Container = styled.View`
  flex: 1;
  padding-top: 30;
  padding-horizontal: 20;
  align-items: center;
`;

const Title = styled.Text`
  font-size: 20;
  margin-vertical: 20px;
`;

const SuccessMessage = styled.Text`
  color: green;
`;

const ErrorMessage = styled.Text`
  color: red;
`;
