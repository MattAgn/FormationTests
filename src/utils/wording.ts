export const wording = {
  subscribe: 'Subscribe to our newsletter !',
  emailPlaceholder: 'Email',
  validateEmail: 'Validate',
  emailSent: 'Email sent !',
  basicError: 'Ooops something went wrong...',
};
