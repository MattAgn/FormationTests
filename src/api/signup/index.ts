import { BaseApi } from '../';
import { IValidateEmailParams } from './types';
import { EMAIL_API_ENDPOINT } from '../config';

export class SignUpApi {
  public static subscribeNewsletter(subscribeNewsletterParams: IValidateEmailParams) {
    return BaseApi.baseCall()
      .url(EMAIL_API_ENDPOINT)
      .post(subscribeNewsletterParams)
      .res();
  }
}
