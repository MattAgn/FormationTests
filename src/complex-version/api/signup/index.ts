import wretch from 'wretch';

import { EMAIL_API_ENDPOINT } from '../config';

interface IValidateEmailParams {
  email: string;
}

export class SignUpApi {
  public static subscribeNewsletter(subscribeNewsletterParams: IValidateEmailParams) {
    return wretch()
      .url(EMAIL_API_ENDPOINT)
      .post(subscribeNewsletterParams)
      .res();
  }
}
