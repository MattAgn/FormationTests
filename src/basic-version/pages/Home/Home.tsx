import React, { useState, FunctionComponent } from 'react';
import styled from 'styled-components/native';
import wretch from 'wretch';

import { Button } from '../../../components/Button';
import { wording } from '../../../utils/wording';
import { EMAIL_API_ENDPOINT } from '../../../complex-version/api/config';
import { Container, Card, Title, Input } from '../../../components/StyledComponents';
import { NavigationScreenProps } from 'react-navigation';
import { Routes } from '../../../complex-version/navigation/routes';

export const Home: FunctionComponent<NavigationScreenProps> = ({ navigation }) => {
  const [email, setEmail] = useState('');
  const [isSuccess, setIsSuccess] = useState<boolean | null>(null);

  const onValidate = () => {
    wretch(EMAIL_API_ENDPOINT)
      .post({ email })
      .res()
      .then(() => setIsSuccess(true))
      .catch(() => setIsSuccess(false));
  };

  const onAbout = () => navigation.navigate(Routes.About);

  return (
    <Container>
      <Card>
        <Title>{wording.subscribe}</Title>
        <Input placeholder={wording.emailPlaceholder} value={email} onChangeText={setEmail} />
        <Button title={wording.validateEmail} onPress={onValidate} />
        <Button title={wording.about} onPress={onAbout} />
        {isSuccess && <SuccessMessage>{wording.subscriptionSuccessful}</SuccessMessage>}
        {isSuccess === false && <ErrorMessage>{wording.basicError}</ErrorMessage>}
      </Card>
    </Container>
  );
};

const SuccessMessage = styled.Text`
  color: green;
`;

const ErrorMessage = styled.Text`
  color: red;
`;
