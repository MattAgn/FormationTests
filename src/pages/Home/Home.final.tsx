import React from 'react';
import styled from 'styled-components/native';
import { Button } from '../../components/Button';
import { useDispatch } from 'react-redux';
import { SignUpActions } from '../../modules/signup/actions';
import { Formik } from 'formik';
import { Input } from '../../components/Input';

interface EmailValue {
  email: string;
}

export const HomeFinal = () => {
  const dispatch = useDispatch();

  const onSubmit = ({ email }: EmailValue) => dispatch(SignUpActions.subscribeNewsletter(email));

  return (
    <Container>
      <Title>Subscribe to our newsletter</Title>
      <Formik
        onSubmit={onSubmit}
        initialValues={{ email: '' }}
        validateOnBlur={false}
        validateOnChange={false}
        render={({ handleSubmit }) => (
          <>
            {/* 
            //@ts-ignore */}
            <Input placeholder="Email" name="email" />
            <Button title="Validate" onPress={handleSubmit} />
          </>
        )}
      />
    </Container>
  );
};

const Container = styled.View`
  flex: 1;
  padding-top: 40;
  padding-horizontal: 20;
  align-items: center;
`;

const Title = styled.Text`
  font-size: 20;
  margin-vertical: 20px;
`;
