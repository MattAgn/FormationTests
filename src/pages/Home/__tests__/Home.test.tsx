import React from 'react';
import 'jest-styled-components';
import { fireEvent, waitForElement, render } from 'react-native-testing-library';
import fetchMock from 'fetch-mock';

import { EMAIL_API_ENDPOINT } from '../../../api/config';
import { Home } from '../Home';
import { wording } from '../../../utils/wording';

describe('[Page] Home', () => {
  beforeEach(() => {
    fetchMock.reset();
  });
  it('should display succesful message on successful subscribe', async () => {
    // Setup
    fetchMock.post(EMAIL_API_ENDPOINT, 200);
    const { getByPlaceholder, queryByText, getByText } = render(<Home />);
    // What the user sees
    const EmailInput = getByPlaceholder(wording.emailPlaceholder);
    const ValidateButton = getByText(wording.validateEmail);
    // What the user does
    fireEvent.changeText(EmailInput, 'hello@bam.co');
    fireEvent.press(ValidateButton);
    // What feedback the user should expect
    const SuccessMessage = await waitForElement(() => queryByText(wording.emailSent));
    expect(SuccessMessage).toBeTruthy();
  });

  it('should display error message on failed subscribe', async () => {
    // Setup
    fetchMock.post(EMAIL_API_ENDPOINT, 400);
    const { getByPlaceholder, queryByText, getByText } = render(<Home />);
    // What the user sees
    const EmailInput = getByPlaceholder(wording.emailPlaceholder);
    const ValidateButton = getByText(wording.validateEmail);
    // What the user does
    fireEvent.changeText(EmailInput, 'hellobam.co');
    fireEvent.press(ValidateButton);
    // What feedback the user will see
    const ErrorMessage = await waitForElement(() => queryByText(wording.basicError));
    expect(ErrorMessage).toBeTruthy();
  });
});
