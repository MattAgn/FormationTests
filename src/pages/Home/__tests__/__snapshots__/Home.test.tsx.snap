// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Page] Home should display error message on failed subscribe 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "paddingHorizontal": 20,
        "paddingTop": 30,
      },
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "fontSize": 20,
          "marginVertical": 20,
        },
      ]
    }
  >
    Subscribe to our newsletter
  </Text>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="Email"
    rejectResponderTermination={true}
    underlineColorAndroid="transparent"
    value="hellobam.co"
  />
  <View
    accessible={true}
    clickable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "green",
            "borderRadius": 6,
            "color": "white",
            "marginVertical": 20,
            "paddingVertical": 10,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "white",
            },
          ]
        }
      >
        Validate
      </Text>
    </View>
  </View>
  <Text
    style={
      Array [
        Object {
          "color": "red",
        },
      ]
    }
  >
    Oups something went wrong...
  </Text>
</View>
`;
